elixir/apps/domain/test/domain/resources_test.exs

```diff
@@ -156,10 +157,73 @@ defmodule Domain.ResourcesTest do
                address: ["can't be blank"],
                name: ["should be at most 255 character(s)"],
                filters: ["is invalid"],
-               connections: ["is invalid"]
+               connections: ["is invalid"],
+               type: ["can't be blank"]
              }
     end
 
+    test "validates dns address", %{subject: subject} do
+      attrs = %{"address" => String.duplicate("a", 256), "type" => "dns"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      assert "should be at most 253 character(s)" in errors_on(changeset).address
+
+      attrs = %{"address" => "a", "type" => "dns"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      refute Map.has_key?(errors_on(changeset), :address)
+    end
+
+    test "validates cidr address", %{subject: subject} do
+      attrs = %{"address" => "192.168.1.256/28", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      assert "is not a valid CIDR range" in errors_on(changeset).address
+
+      attrs = %{"address" => "192.168.1.1", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      assert "is not a valid CIDR range" in errors_on(changeset).address
+
+      attrs = %{"address" => "100.64.0.0/8", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      assert "can not be in the CIDR 100.64.0.0/10" in errors_on(changeset).address
+
+      attrs = %{"address" => "fd00:2011:1111::/102", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      assert "can not be in the CIDR fd00:2011:1111::/106" in errors_on(changeset).address
+
+      attrs = %{"address" => "::/0", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      refute Map.has_key?(errors_on(changeset), :address)
+
+      attrs = %{"address" => "0.0.0.0/0", "type" => "cidr"}
+      assert {:error, changeset} = create_resource(attrs, subject)
+      refute Map.has_key?(errors_on(changeset), :address)
```

Comment:
Hmm, may want to allow these for full-tunnel configurations, but can cross that later.